nRF2 clock control API

bjarki-andreasen opened the pull request:
This PR continues from #76430

anangl committed:
dts: nordic: nrf54h20: Add references to clock controllers

Add references to clock controllers so users can obtain
those for particular peripherals with DT macros like
DT_CLOCK_CTRL()

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl committed:
drivers: clock_control: Add support for nRF54H20 clock controllers

Add custom clock_control API for nRF platforms that allows requesting
clocks with specified minimal required attributes (accuracy, precision,
and frequency). Provide an implementation of this API for FLL16M, HFXO,
HSFLL, and LFCLK controllers in the nRF54H20 SoC.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl committed:
samples: boards: nrf: Add nrf2_clock_control sample

Add a sample that presents how the custom nrf2_clock_control API
introduced for nRF54H20 can be used for the various clocks which
are supported by it.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl committed:
dts: nordic: nrf54h20: Add nodes representing HFXO and LFXO

The HFXO and LFXO oscillators have properties which need to be
specified in the devicetree. These properties are used by clock
controllers, which get these properties from the devicetree.

The BICR also contains these properties. For now, the properties
are duplicated in the bicr node in the devicetree.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl committed:
boards: nrf54h20dk: Add parameters of mounted HFXO and LFXO

These parameters will be used by the clock controller drivers.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: bindings: add nrf-fll16m and nrf-lfclk bindings

Add bindings for nrf-fll16m and nrf-lfclk and update in-tree nodes
and boards to use them in place of the fixed-clock binding.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

hubertmis commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -0,0 +1,256 @@
+	uint16_t accuracy : 15;
+	uint16_t precision : 1;
+	nrfs_clock_src_t src;
+} clock_options[] = {
```
It looks like this array is sorted by accuracy (from the worst being the highest ppm number). If that's a requirement for the algorithm, it would be good to note it somewhere, maybe in a comment?

hubertmis commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -192,6 +204,28 @@ static int api_get_rate_hsfll(const struct device *dev,
+	nrf_hsfll_trim_set(dev_config->hsfll, &trim);
+
+	nrf_hsfll_task_trigger(dev_config->hsfll, NRF_HSFLL_TASK_FREQ_CHANGE);
+	/* HSFLL task frequency change needs to be triggered twice to take effect.*/
```
You could refer to HMPAN-23 here for traceability.

hubertmis reviewed (commented):
We should agree on dt bindings. I would like to remove redundancy from there. Other than that, it looks good to go.

hubertmis commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -192,6 +204,28 @@ static int api_get_rate_hsfll(const struct device *dev,
 {
+	const struct hsfll_dev_config *dev_config = dev->config;
+	nrf_hsfll_trim_t trim = {
+		.vsup   = sys_read32(dev_config->trim_vsup),
+		.coarse = sys_read32(dev_config->trim_coarse),
+		.fine   = sys_read32(dev_config->trim_fine),
```
Settint TCOEF trim is still missing. Are we aiming to fix it in this PR? If not we should create a ticket to make sure we don't forget.

bjarki-andreasen replied:
Created a ticket, this PR just moves the existing init code from SOC to the clock driver, so it arguably is out of scope :)

nordic-bami committed:
tests: drivers: Implement nrf2 clock control api tests

Implement tests for the NRF2 clock control API

Signed-off-by: Bartosz Miller <bartosz.miller@nordicsemi.no>

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_common.h:
```diff
@@ -0,0 +1,89 @@
+/**
+ * @brief Request or release lrcconf main power domain
+ */
+void nrf2_clock_request_lrcconf_poweron_main(struct nrf2_clock_lrcconf_sink *sink);
+void nrf2_clock_release_lrcconf_poweron_main(struct nrf2_clock_lrcconf_sink *sink);
```
agreed, this is a temporary solution :)

hubertmis commented on dts/bindings/clock/nordic,nrf-fll16m.yaml:
```diff
@@ -0,0 +1,34 @@
+    type: phandle
+    description: |
+      The LFXO clock is used as a clock source if the clock source is closed-loop and
+      the LFXO clock is not available.
```
typo: shoud be "LFXO clock is available"

bjarki-andreasen replied:
fixed

hubertmis commented on dts/bindings/clock/nordic,nrf-lfclk.yaml:
```diff
@@ -0,0 +1,32 @@
+  hfxo-clock:
+    type: phandle
+    description: |
+      The HFXO clock is used as a clock source if the SYNTH clock source is selected
+      and the LFXO clock is not available.
+
+  lfxo-clock:
+    type: phandle
+    description: |
+      The LFXO clock is used as a clock source if the SYNTH clock source is selected
+      and the LFXO clock is available.
```
You could add a note to both that these clocks are used indirectly through FLL16M in BYPASS mode. Something like:

The xxXO clock is used indirectly as a clock source, through FLL16M in the BYPASS mode, if the SYNTH clock source is selected...

Or maybe easier to understand:

The xxXO clock is used as a clock source if...
The xxXO is used indirectly. It is a clock source for FLL16M configured in a BYPASS mode which is a direct clock source for LFCLK in the SYNTH mode.

bjarki-andreasen replied:
updated to mention indirect use through FLL16M

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -0,0 +1,261 @@
+		.mode = FLL16M_MODE_OPEN_LOOP,
+	},
+	{
+		/* Closed-loop mode uses LFXO as source if present, HFXO otherwise */
```
It seems this comment applies to lines 36-40 now

bjarki-andreasen replied:
moved to line 36

hubertmis reviewed (dismissed):
Looks good. A little nitpicking around the doc strings and comments. But for me, it looks good to be merged and mistakes can be fixed later.

hubertmis commented on dts/bindings/clock/nordic,nrf-fll16m.yaml:
```diff
@@ -0,0 +1,34 @@
+  hfxo-clock:
+    type: phandle
+    description: |
+      The HFXO clock is used as a clock source if the clock source is bypass, and if
```
I missed this during the previous review iteration. Looks like another typo:
```suggestion
      The HFXO clock is used as a clock source if the clock source is bypass, or if
```

bjarki-andreasen replied:
done :)

hubertmis reviewed (dismissed):
No more comments :)

gmarull commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -25,20 +25,23 @@
 			reg = <3>;
 			device_type = "cpu";
+			clocks = <&cpurad_hsfll>;
 			clock-frequency = <DT_FREQ_M(256)>;
```
we should at some point remove clock-frequency as it has become redundant

bjarki-andreasen replied:
outside the scope of this PR

gmarull commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi:
```diff
@@ -21,3 +21,19 @@
+	status = "okay";
+	open-loop-accuracy-ppm = <20000>;
+	closed-loop-base-accuracy-ppm = <5000>;
+	hfxo-clock = <&hfxo>;
+	lfxo-clock = <&lfxo>;
```
why is this modified at board level?

gmarull replied:
also, this should be:

```
clocks = <&hfxo>, <&lfxo>;
clock-names = "hfxo", "lfxo";
```

bjarki-andreasen replied:
Changed to use clocks property, and moved to the common nrf54h20.dtsi

gmarull commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi:
```diff
@@ -21,3 +21,19 @@
+	status = "okay";
+	lfrc-accuracy-ppm = <500>;
+	lflprc-accuracy-ppm = <1000>;
+	hfxo-clock = <&hfxo>;
+	lfxo-clock = <&lfxo>;
```
same

gmarull commented on drivers/clock_control/clock_control_nrf2_common.c:
```diff
@@ -0,0 +1,197 @@
+#include "clock_control_nrf2_common.h"
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+LOG_MODULE_REGISTER(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
```
move after includes

bjarki-andreasen replied:
Moved after includes

gmarull commented on samples/boards/nrf/nrf2_clock_control/src/main.c:
```diff
@@ -0,0 +1,98 @@
+	ret = nrf_clock_control_release(sample_clock_dev, &spec);
+	if (ret < 0) {
+		printk("failed to release requested clock specs\n");
+		return ret;
```
main must always return 0

bjarki-andreasen replied:
updated to return 0

bjarki-andreasen commented:
Since we are now using the clocks property, the property descriptions for the properties `hfxo_clock` and `lfxo_clock` has been moved to the the descriptions of the bindings, along with a dts example for both `fll16m` and `lfclk`

bjarki-andreasen commented:
missed a couple of log register macros not placed after header :)

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -0,0 +1,263 @@
+#define FLL16M_CLOSED_LOOP_ACCURACY (FLL16M_CLOSED_LOOP_BASE_ACCURACY + FLL16M_HFXO_ACCURACY)
+#endif
+
+/* Clock options sorted from lowest to highest accuracy */
```
Now when I'm thinking about this array, it actually is sorted from the lowest to the highest power consumption.
In the case of this specific array, accuracy is also improving. But it's not necessarily the rule for all similar arrays.

We aim to sort these arrays by power consumption to make the algorithm select the mode with _the lowest power consumption_, which meets the requested accuracy and precision numbers.

bjarki-andreasen replied:
I will leave this for now, since the API clock spec requests set frequency, accuracy and precision. That's why these arrays focus on those properties for ranking

hubertmis reviewed (dismissed):
When reading the PR again a few more text improvements came to my mind

bjarki-andreasen commented on dts/bindings/clock/nordic,nrf-fll16m.yaml:
```diff
@@ -0,0 +1,42 @@
+  The FLL16M can use the following clocks as clock sources:
+
+    - HFXO: The HFXO clock is used as a clock source if the
+      clock source is bypass, or if the clock source is
```
updated to use mode for the clocks FLL16M and LFCLK and clock source for the clocks HFXO and LFXO

hubertmis commented on dts/bindings/clock/nordic,nrf54h-hfxo.yaml:
```diff
@@ -0,0 +1,30 @@
+
+  startup-time-us:
+    type: int
+    description: Clock startup time in micro seconds
```
The BICR generator should take this value as an input.
@hakonfam, FYI

bjarki-andreasen commented:
Added a DNM until I have added some patches we identified downstream :)

bjarki-andreasen commented:
The latest changes modify Kconfig to not select NRF2_CLOCK_CONTROL by default, and removes the commit moving trimming of hsfll clock from the device driver init to the soc to ensure it is executed even if the device driver is not included (included by NRF2_CLOCK_CONTROL). This helps save flash in resource constrained applications.

hubertmis commented:
> The latest changes modify Kconfig to not select NRF2_CLOCK_CONTROL by default, and removes the commit moving trimming of hsfll clock from the device driver init to the soc to ensure it is executed even if the device driver is not included (included by NRF2_CLOCK_CONTROL). This helps save flash in resource constrained applications.

Is it temporary or a final solution? I understood we aim to follow other vendors and keep clock-related initialization in the clock drivers. If we aim for resource-constrained applications, we should introduce Kconfigs to remove parts of the clock drivers. However, initialization should be obligatory if it's required for proper SoC operations.

bjarki-andreasen commented:
> > The latest changes modify Kconfig to not select NRF2_CLOCK_CONTROL by default, and removes the commit moving trimming of hsfll clock from the device driver init to the soc to ensure it is executed even if the device driver is not included (included by NRF2_CLOCK_CONTROL). This helps save flash in resource constrained applications.
> 
> Is it temporary or a final solution? I understood we aim to follow other vendors and keep clock-related initialization in the clock drivers. If we aim for resource-constrained applications, we should introduce Kconfigs to remove parts of the clock drivers. However, initialization should be obligatory if it's required for proper SoC operations.

It is temporary :) clock init should definitely be moved to the clock driver once we figure out how optimize it for size :) I have created an internal ticket to track it :)

anangl commented on drivers/clock_control/clock_control_nrf2_common.h:
```diff
@@ -0,0 +1,89 @@
+ * @param _onoff_cnt number of clock configuration options to be handled;
+ *                   for each one a separate onoff manager instance is used.
+ */
+#define NRF2_STRUCT_CLOCK_CONFIG(type, _onoff_cnt) \
```
This is not a public API, so I don't think we need to use these `nrf2_` prefixes here (which IMO look a bit weird without `clock_control_` before them).

bjarki-andreasen replied:
done

bjarki-andreasen commented:
Removed `nrf2_` namespace from private functions of the nrf2 clock control drivers, and adjusted licenses I had missed from `LicenseRef-Nordic-5-Clause` to `Apache-2.0`

bjarki-andreasen commented on dts/bindings/clock/nordic,nrf-lfclk.yaml:
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
done

bjarki-andreasen commented on dts/bindings/clock/nordic,nrf-fll16m.yaml:
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
done

anangl commented on drivers/clock_control/clock_control_nrf2_hsfll.c:
```diff
@@ -0,0 +1,237 @@
+	}
+
+	frequency = spec->frequency == NRF_CLOCK_CONTROL_FREQUENCY_MAX
+		  ? MHZ(320)
```
This value is quite hidden here. It would be better to have it defined somewhere near the `clock_options` array.

bjarki-andreasen replied:
added defines
```
#define HSFLL_FREQ_LOW    MHZ(64)
#define HSFLL_FREQ_MEDLOW MHZ(128)
#define HSFLL_FREQ_HIGH   MHZ(320)
```
and used this across file

anangl commented on samples/boards/nrf/nrf2_clock_control/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config SAMPLE_CLOCK_FREQUENCY_HZ
+	int "Frequency specification to request from clock in Hz"
+
+config SAMPLE_CLOCK_ACCURACY_PPM
+	int "Accuracy specification to request from clock in PPM"
+
+config SAMPLE_CLOCK_PRECISION
+	int "Precision specification to request from clock"
```
Wouldn't it make sense to use `default 0` for these options (the API defines that 0 means: no specific requirement) and change them only when a certain value is required?

bjarki-andreasen replied:
added default 0 along with help to indicate what values of 0, 1 and >1 means

anangl commented on samples/boards/nrf/nrf2_clock_control/configs/fll16m.conf:
```diff
@@ -0,0 +1,8 @@
+
+CONFIG_CLOCK_CONTROL_NRF2=y
+
+CONFIG_SAMPLE_CLOCK_FREQUENCY_HZ=16000000
```
Why request any value here when there is anyway no other choice than 16 MHz? Similarly for `lfclk`, it is always 32768. And for UART, users don't actually need to know/care what is the frequency of the clock that drives it.

bjarki-andreasen replied:
removed frequency (now uses default 0)

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_common.c:
```diff
@@ -0,0 +1,198 @@
+ */
+
+#include "clock_control_nrf2_common.h"
+#include <zephyr/kernel.h>
```
removed #include <zephyr/kernel.h>

bjarki-andreasen commented on samples/boards/nrf/nrf2_clock_control/src/main.c:
```diff
@@ -0,0 +1,98 @@
+
+static K_SEM_DEFINE(sample_sem, 0, 1);
+
+static void sample_notify_cb(void)
```
updated signature

anangl commented on samples/boards/nrf/nrf2_clock_control/CMakeLists.txt:
`nrf2_clock_control` does not seem to be a good name for this sample. I'd go with just `clock_control`.

bjarki-andreasen replied:
renamed to clock_control

anangl commented on tests/drivers/clock_control/nrf2_clock_control/CMakeLists.txt:
Similarly here, instead of `nrf2_clock_control`, I'd use `nrf_clock_control`.

bjarki-andreasen replied:
renamed to nrf_clock_control

bjarki-andreasen commented on samples/boards/nrf/nrf2_clock_control/Kconfig:
```diff
@@ -0,0 +1,38 @@
+	help
+	  0  -> ignore frequency
+	  1  -> use max frequency
+	  >1 -> use at minimum selected frequency
```
Updated to state:
```
	  >1 -> use at minimum selected frequency. To select the
	        highest supported frequency use UINT32_MAX.
```

anangl replied:
It should say `>0`, actually.

bjarki-andreasen replied:
Updated to >0

hubertmis commented:
Temporarily, we should not use closed-loop modes. Should we update this PR or remove the closed-loop later?

bjarki-andreasen commented:
> Temporarily, we should not use closed-loop modes. Should we update this PR or remove the closed-loop later?

I would like to remove closed loop later, to keep consistency between the commits in this PR and the downstream PR (which is already merged)
